### DISCUSSION TITLE
7.0.0

### DISCUSSION THREAD

**Discussion by Ghost-chu** [PBH-BTN/PeerBanHelper#655] [公告]
经过一段时间的沉淀，我们很高兴宣布 v7.0.0 版本现已正式发布~  

## 2000 Stars ⭐ 庆贺

在开始之前，请允许我感谢各位朋友们一直以来的支持，我谨代表 PBH-BTN 向各位致以诚挚的谢意 :)  
在过去的一段时间内，我们的项目受到了大家的关注和热烈欢迎，这是我们远远未曾设想过的。PeerBanHelper 能发展到今天离不开大家的积极支持和鼓励，以及非常感谢在 Issues 上打开议题反馈错误的朋友们，帮助 PeerBanHelper 变得越来越好。    

同时，我们也很感谢一直以来通过加入 BTN 网络来为反吸血贡献力量的伙伴们。在你们的支持下，自 2024 年 8 月 20 日以来，连接到 BTN 的 PBH 共拦截和封禁 Peer 超过 **6,842,412** 次。其中，通过 BTN 用户贡献的规则识别并拦截了来自 **228,700** 个 IP 地址的 **3,323,527** 次恶意刷流攻击。

这并不是完整的数据，因为还有更多的人未加入 BTN 网络，这些客户端和使用 IP 规则的拦截数据并未计入其中。但我相信，你们的贡献已经帮助到比这更多更多的人。  

尽管关注 BitTorrent 技术的人越来越少。但我们坚信 BitTorrent 不可替代，它是互联网精神的体现：无私分享，不求利益，不求回报。

希望互联网精神能够延续下去，这是我们共同的心愿。共勉！

---

## WebUI 焕然一新

忘记编辑配置文件的麻烦吧！通过 WebUI 直接编辑 PeerBanHelper 的首选项和各个反吸血模块的设置，所见即所得。

| PeerID 封禁 |反吸血模块合影 | 多拨封禁 | 进度作弊检查  |
| ----------- | ------------ | ---------- | --------- |
| ![1](https://github.com/user-attachments/assets/c5a2b5b3-bebb-4829-91ad-a89c1858e99c) | ![image](https://github.com/user-attachments/assets/737ecaac-803d-4723-806d-3b6cf5c48998) | ![3](https://github.com/user-attachments/assets/3420b555-dc94-438f-971f-3d2979cd74a2) | ![2](https://github.com/user-attachments/assets/64e3b276-f1dd-4353-9285-f7a22ec7bf4b) |

## 运行状态 随时掌握

在 WebUI 上直接查看程序状态、控制台日志和 BTN 网络连接状态，不再黑盒。

| BTN 状态 | 运行状态 |
| ------- | -------- |
| ![s2](https://github.com/user-attachments/assets/7a4678ee-6f81-42e6-9163-23f9e0bd3f2e) | ![s1](https://github.com/user-attachments/assets/193aba3d-4f8e-4d95-9c06-c1cb05a28977) |

## 及时跟进 重要信息

新的 PBH 引入了 “警报” 消息推送中心系统。您可以在 WebUI 查看当前活跃的各类警报，亦可以通过电子邮件、Server酱、推送加（提供钉钉、微信公众号、企业微信、短信等渠道）和 Telegram 等方式发送至您的设备。

您还可以在 Active Monitoring 主动监测模块中设置每日流量告警阈值，避免不小心被吸干的情况发生，及时止损。

> [!WARNING]
> 注意！由于 API 限制，每日流量告警阈值功能当前仅对 qBittorrent、qBittorrent EE、Deluge 和 BiglyBT 有效

| 警报中心 | 邮件推送 |
| ------- | ------- |
| ![a1](https://github.com/user-attachments/assets/804b3182-7c41-4255-a0c4-0002ab43db63) | ![a2](https://github.com/user-attachments/assets/80df195b-5cd3-45b2-89c9-db7369036f94) 

## BTN 能力再升级

新增两种 BTN 协议：

* 例外：BTN 侧的白名单，处于白名单内的 IP 不会被封禁；如果已被封禁将会自动解除封禁。这有助于我们的数据遭到污染后的去污染能力
* Peers 历史：提交自启动或上次提交以来的所有 Peers 记录，以便 BTN 服务器分析统计在 BTN 网络上的恶意活动

实践：

通过 Peers 历史提交能力，我们在所有使用 v7.0.0 测试版用户的测试群体内进行了数据分析，在新的 BTN 能力加持下，我们成功捕获了多个从不同用户处反复重复下载文件的恶意 IP 地址。  

注：相同 torrent 值为相同种子；total_uploaded 和 size 的单位是 bytes；upload_percentage 的单位是百分比，100% = 下载了种子 1 次。

![image](https://github.com/user-attachments/assets/9dcfa69b-a7f7-4441-915e-559f61ba7452)

## 生态扩展 支持广泛

新版本的 PeerBanHelper 中更新了对 BitComet 2.11 (20241022) 的 WebAPI 的支持。  
除此之外，我们将 PeerBanHelper 的平台支持扩展到了：

* macOS 的 Intel / Apple Silicon 设备
* 运行 Linux 系统的 risc-v, ppp64le, 和 loongarch64 设备

其中，由于龙芯处理器平台的生态缺失，暂无 Docker 镜像可用。对于其它支持的平台，可下载安装程序或者拉取对应架构的 Docker 镜像使用。

至此，在 v7.0.0 版本中，我们已正式支持：

* Windows x86_64
* Windows arm64
* macOS Intel/Apple Silicon
* Linux x86_64/aarch64/riscv64/ppp64le/loongarch64
* Linux (musl) x86_64/aarch64
* FreeBSD x86_64/arm64

您现在可以在几乎所有主流平台上运行 PeerBanHelper。

---

## 新增平台适配

* 核心适配现在支持 loongarch64 （龙芯处理器 - 新世界）平台，感谢群友 lang @Gaojianli @Ghost-chu @paulzzh 
* 核心适配现在支持 riscv64 平台，感谢群友 lang @Ghost-chu 
* 核心适配现在“理论”支持 ppc64le 平台，欢迎各位持有对应平台设备的朋友进行测试 @Ghost-chu 

其中，`riscv64` 和 `ppc64le` 平台已添加 Docker 和安装程序支持。`loongarch64` 由于缺少基础设施，目前只能通过 JAR 手动部署的方式安装，我们也欢迎您贡献 loongarch64 的 Docker 镜像打包方案。

## 新功能

* WebUI 设置界面 @Gaojianli @paulzzh 
  * 通过 WebUI 的设置界面，管理和修改 PeerBanHelper 的首选项和反吸血模块设定
* WebUI 支持查看实时日志 @Gaojianli 
* 新增告警功能，可在 WebUI 显示需要注意的提示，并通过推送渠道推送通知 @Ghost-chu @paulzzh 
  * 支持下列推送渠道：
    * SMTP 邮件发信
    * 推送加
    * Server 酱
    * Telegram
* 新增适配 macOS 系统 @Ghost-chu @paulzzh 
* 新增 BTN Peers 累积流量数据上报协议 @Ghost-chu 
* 新增 BTN 例外规则协议 @Ghost-chu 
* 新增 BTN 状态页，可显示 PBH 连接到 BTN 的情况 @Gaojianli @Ghost-chu 
* 新增流量超限告警，可设置每日上传告警阈值，超限后通过推送渠道发送通知 @Ghost-chu 
* 新增下载器登录故障告警，如果下载器可连接但无法正确登录或返回错误响应，达到一定次数后，将发送告警提示信息 @Ghost-chu 
* 新增内存使用感知，当 PeerBanHelper 的内存即将耗尽且无法释放足够空间时，将发送告警提示信息 @Ghost-chu 

## 更改

* 适配 BitComet 2.11 (20241022) 更新的 WebAPI @Ghost-chu 
  * 适配新的解除封禁 API 接口，快速 PCB 测试现在在 BitComet 上可用
  * 改进 BitComet 的是否登录检查的性能，大幅降低请求数
  * 限制 BitComet 最大并发数到 4，避免连接数过多导致 BitComet 的 HTTP Server 关闭部分连接引发查询错误
* 将部分提示和错误信息本地化 @Ghost-chu 
* 多拨追猎支持为 v4 和 v6 地址分别设置阈值 @Ghost-chu 

## 错误修复

* 修复 Windows 资源管理器重启后无法再与托盘图标交互的问题 @paulzzh 
* 修复如果下载器未就绪，则 PeerBanHelper 启动时更新封禁列表会出错的问题 @Ghost-chu 
* 修复安装器版本 PeerBanHelper 无法正确生成和保存日志文件的问题 @Ghost-chu 
* 修复 BTN 规则更新时出现的空指针报错的问题，此错误并不影响使用 @Ghost-chu 
* 修复处理多下载器时可能在一个 BanWave 中封禁同一个 Peer 多次（于不同下载器），从而触发 DUPLICATE_BAN 提示并触发强制全量同步的问题 @Ghost-chu 
* 修复非 Windows 平台上的系统语言检测问题 @Ghost-chu 
* 修复日志窗口会随机调到最顶部的问题 @Ghost-chu 
* 修复日志窗口每次新日志插入时都跳到最下面的问题 @Ghost-chu 
* 修复 BitComet 上可能意外解析 IP 地址为域名并引发错误的问题 @Ghost-chu 
* 修复了执行重载操作时，程序代理服务器设置未正确重载的问题 @Ghost-chu 
* 修复了执行重载操作时，没有从封禁列表中自动解除已在白名单地址段的 IP 地址 @Ghost-chu 
* 修复了 BTN 网络规则模块的 peer_id 以及 port 的规则判断检查不生效的问题 @Ghost-chu 
* 修复 UNBAN_IP 检查条件错误反转的问题 @Ghost-chu 

## 删除的功能

* 删除了 JavaFx GUI @Ghost-chu 
  * 集编程困难、依赖繁琐、容易崩溃等多个问题于一体的 UI 框架，我们回到了 Swing 的正路上，并节约了 12MB 的磁盘空间
  * 由于现在只使用 Swing UI，因此 GUI Safe-Mode (Swing) 启动方式已被移除
* 运行时依赖下载和加载工具 @Ghost-chu
  * 在过去的多版本中，证明在中国大陆的网络环境下，在运行时解决依赖不是一种好的方式
  * 通过更改打包方式，现在所有依赖将一次性全部打包，全量分发，不再在运行时处理
* 移除规则命中率统计页面 @paulzzh 

## 内部更改

* 更改了打包方式，不再输出 uber jar，而是通过 -cp 加载 jars 并引导主类启动 @Ghost-chu @paulzzh 
  * 解决了因重定位 jar 导致破坏数字签名，并导致编译失败/运行时类加载失败的问题
* 在执行重载过程时，解除位于白名单 IP 段配置的 IP 地址的封禁 @Ghost-chu 
* 从 takari 换回传统编译 @Ghost-chu 
* 使用 logback 代替 log4j2 @Ghost-chu 
* 各类性能优化和代码清理

## Docker

DockerHub: `ghostchu/peerbanhelper:v7.0.0`
阿里云国内镜像加速: `registry.cn-hangzhou.aliyuncs.com/ghostchu/peerbanhelper:v7.0.0`

---

## 公益 Tracker 项目正在公测

由 PBH-BTN 维护的 Sparkle Tracker 现在正在公测。这是一个高性能的公共 Tracker 服务，除提供基本 Tracker 功能外，它还和 BTN 进行了整合。通过结合两者的数据，我们可以更加了解当前对 BT 网络的攻击态势。  

您可以将下面的 Tracker 地址添加到您的下载器的 Tracker 列表中使用。

```
https://sparkle.ghostchu-services.top/tracker/announce
```

## 鸣谢

* 感谢 [大河李斯特视频工作室](https://space.bilibili.com/12562485) 老师制作的 [【看番教程39】怎么样从网上下载资源？磁链/种子到底是啥？？评论区神秘代码要怎么用？什么是BT协议？BT入门保姆级教程，webUI、屏蔽吸血客户端PCDN详解](https://www.bilibili.com/video/BV1VyyUYVEMq) 让更多人了解磁力链接和 BT 种子背后的故事和技术，降低了新人踏入 BitTorrent 世界的门槛，也让更多人知道如何保护自己的下载器免受恶意攻击的侵害。对于一个较小和相对固定的社区来说，入门门槛高是一个屡见不鲜的问题。感谢您制作的系列视频降低了新人入门的门槛，也让更多人了解了这些网络技术背后的所发生的事情。请允许我们对您的无私奉献致以敬意
* 感谢 [我不是矿神](https://imnks.com/) 在过去一段时间内对 PeerBanHelper 项目的支持，帮助在群晖 NAS 平台上的用户指导部署 PBH，以及在 SPK 软件源中分发 PeerBanHelper。NAS 用户因长期开机做种，更易成为吸血鬼攻击的目标，有时一不小心，可能一觉睡醒天就塌了。IMNKS 在 NAS 用户群体中的宣传让更多的 NAS 用户了解到了 BT 刷流吸血的存在，并避免了更多用户深受其害
* 以及其他默默在背后支持我们的朋友们 :)

<hr /><em>This discussion was created from the release <a href='https://github.com/PBH-BTN/PeerBanHelper/releases/tag/v7.0.0'>7.0.0</a>.</em>

**Comment by Tigerest** [comment]
恭喜两千star！

**Comment by poplar-at-twilight** [comment]
与 `PeerBanHelper.jar` 一起发布的 `libraries.tar.gz` 会经常更新吗？

**Reply by Ghost-chu** [reply]
会的

**Reply by poplar-at-twilight** [reply]
好的，我再去把脚本改改👏😂

**Comment by feng1m8** [comment]
windows便携版本的压缩文件好像损坏了😿

**Comment by TomoeMami** [comment]
```
riko@riko:~/PBH$ java -jar -Xmx512M -XX:+UseZGC -XX:+UseStringDeduplication -XX:+ShrinkHeapInSteps -jar PeerBanHelper.jar

Exception in thread "main" java.lang.NoClassDefFoundError: org/bspfsystems/yamlconfiguration/configuration/InvalidConfigurationException
        at com.ghostchu.peerbanhelper.MainJumpLoader.main(MainJumpLoader.java:17)
Caused by: java.lang.ClassNotFoundException: org.bspfsystems.yamlconfiguration.configuration.InvalidConfigurationException
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
        ... 1 more
```

编辑：已从jar文件换为docker，正常使用了。
